internal/pkgs/wifimanager/controller.go

```diff
@@ -4,101 +4,444 @@ package wifimanager
+	var output []*wifi_api.AccessPoint
 
 	// Input validation
 	if ctx == nil {
-		return parsedOutput, fmt.Errorf("context cannot be nil")
+		return output, fmt.Errorf("context cannot be nil")
 	}
 
-	nmcliRunCmd := "/run/current-system/sw/bin/nmcli"
-	nmcliRunCmd += " -f IN-USE,SSID,SIGNAL,SECURITY "
-	nmcliRunCmd += " device wifi "
+	// Iterate over Wi-Fi devices
+	for _, device := range c.wifiDevices {
+		var apPaths []dbus.ObjectPath
 
-	// Get wifi list from nmcli
-	cmd := exec.Command("/bin/sh", "-c", nmcliRunCmd)
-	networks, err := cmd.Output()
+		err := device.Call("org.freedesktop.NetworkManager.Device.Wireless.GetAllAccessPoints", 0).Store(&apPaths)
+		if err != nil {
+			return output, fmt.Errorf("failed to get access points: %s", err)
+		}
+		// Iterate over access points and append into output
+		for _, apPath := range apPaths {
+			var accesspoint AP
 
-	if err != nil {
-		return parsedOutput, fmt.Errorf("error starting application: %s (%s)", "nmcli", err)
+			ap := c.conn.Object("org.freedesktop.NetworkManager", apPath)
+			accesspoint, err := GetAPData(ap)
+			if err != nil {
+				return output, fmt.Errorf("failed to get accesspoint data: %s", err)
+			}
+			accesspoint.Security = GetAPSecurity(accesspoint)
+
+			// Append variables into output
+			network := wifi_api.AccessPoint{
+				SSID:     accesspoint.SSID,
+				Signal:   accesspoint.Strength,
+				Security: accesspoint.Security,
+			}
+			output = append(output, &network)
+		}
 	}
+	return output, nil
+}
 
-	parsedOutput["IN-USE"] = []string{}
-	parsedOutput["SSID"] = []string{}
-	parsedOutput["SIGNAL"] = []string{}
-	parsedOutput["SECURITY"] = []string{}
+func (c *WifiController) GetActiveConnection(ctx context.Context) (string, uint32, string, error) {
+	var accesspoint AP
 
-	exp := regexp.MustCompile(`(?m)^(\*?)\s+(.*\S)\s+([0-9]+)\s+(\w+\s?\w+|--)([\s]+)?$`)
-	find := exp.FindAllSubmatch(networks, -1)
-	for i := range find {
-		parsedOutput["IN-USE"] = append(parsedOutput["IN-USE"], string(find[i][1]))
-		parsedOutput["SSID"] = append(parsedOutput["SSID"], string(find[i][2]))
-		parsedOutput["SIGNAL"] = append(parsedOutput["SIGNAL"], string(find[i][3]))
-		parsedOutput["SECURITY"] = append(parsedOutput["SECURITY"], string(find[i][4]))
+	// Input validation
+	if ctx == nil {
+		return "", 0, "", fmt.Errorf("context cannot be nil")
 	}
 
-	return parsedOutput, nil
+	// Iterate over Wi-Fi devices
+	for _, device := range c.wifiDevices {
+		var ap string
+		activeAPPath, err := device.GetProperty("org.freedesktop.NetworkManager.Device.Wireless.ActiveAccessPoint")
+		if err != nil {
+			return "", 0, "", fmt.Errorf("failed to get active access point path: %s", err)
+		}
+		activeAPPath.Store(&ap)
+
+		activeAP := c.conn.Object("org.freedesktop.NetworkManager", dbus.ObjectPath(ap))
+		accesspoint, err = GetAPData(activeAP)
+		if err != nil {
+			return "", 0, "", fmt.Errorf("failed to get access point data: %s", err)
```

Comment:
Should there be a better response for no active connection than an error?

(Also, this will behave badly if there are multiple wifi devices, but that's not a supported use-case)

Reply:
Active connection boolean is added instead of the error.